java-checks/src/main/java/org/sonar/java/symexecengine/SymbolicExecutionCheck.java

```diff
@@ -0,0 +1,82 @@
+/*
+ * SonarQube Java
+ * Copyright (C) 2012 SonarSource
+ * dev@sonar.codehaus.org
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02
+ */
+package org.sonar.java.symexecengine;
+
+import org.sonar.plugins.java.api.semantic.Symbol;
+import org.sonar.plugins.java.api.tree.ExpressionTree;
+import org.sonar.plugins.java.api.tree.MethodTree;
+import org.sonar.plugins.java.api.tree.Tree;
+import org.sonar.plugins.java.api.tree.VariableTree;
+
+import java.util.List;
+
+public abstract class SymbolicExecutionCheck {
+
+  protected abstract boolean isSymbolRelevant(Symbol symbol);
+
+  /**
+   * called prior analysis of the given method.
+   *
+   * @param executionState execution state
+   * @param tree method to be analyzed
+   */
+  protected void initialize(ExecutionState executionState, MethodTree tree) {
+  }
+
+  /**
+   * called when an assignment expression or a variable declaration with an initializer is evaluated.
+   *
+   * @param executionState current execution state
+   * @param tree underlying tree
+   * @param variable variable symbol
+   * @param expression expression
+   */
+  protected void onAssignment(ExecutionState executionState, Tree tree, Symbol variable, ExpressionTree expression) {
+  }
+
+  /**
+   * called when a constructor or method is invoked.
+   *
+   * @param executionState execution state
+   * @param tree method invocation, constructor tree or new class tree
+   */
+  protected void onInvocation(ExecutionState executionState, Tree tree, List<ExpressionTree> arguments) {
```

Comment:
Could be a method invocation or a new class.

As below this should be used only for issue reporting.